# Комментарий наставника
Привет! 

Я не стал удалять старые комментарии. Если ты посчитаешь нужным - можешь их удалить, чтобы не мешали. 

Как я и писал раньше - хорошая и очень чистая, последовательная работа. У меня есть какое-то количество замечаний, но не будет ничего страшного если они перекочуют в следующий проект и мы их поправим уже там. 

Если очень коротко советы/замечания/предложения:
1. можно посмотреть в сторону метода transform и fillna (https://stackoverflow.com/questions/19966018/pandas-filling-missing-values-by-mean-in-each-)
2. Еще для деления непрерывной величины на группы есть методы qcut.


Приступай к следующему проекту! По поводу причин отставания: надеюсь все и дальше будет хорошо. Не бойся отправлять проект, если ты не уверен в каком-то одном пункте. Для этого и нужен наставник, чтобы что-то поправить и/или подсказать. Это не повод отсылать совсем сырую работу, но и сидеть несколько дней в ступоре тоже не нужно.

# Комментарий наставника
Лучше начинать с короткого описания, что вообще собственно тут происходит. Обычно,  конечно, ноутбуки существуют в каком-то контексте: приложены к письму или какой-то задаче в соответствующей системе, но возращаться к работе проще, если есть короткое описание о чем оно.

# Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку. 

 Описание данных:
 
     children — количество детей в семье
     days_employed — трудовой стаж в днях
     dob_years — возраст клиента в годах
     education — образование клиента
     education_id — идентификатор образования
     family_status — семейное положение
     family_status_id — идентификатор семейного положения
     gender — пол клиента
     income_type — тип занятости
     debt — имел ли задолженность по возврату кредитов
     total_income — доход в месяц
     purpose — цель получения кредита

# Задание 1. 
Проведите исследование

## Шаг 1. 
Откройте файл с данными и изучите общую информацию.

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')


data.to_csv('data.csv')

#вывод первых 11 строк набора данных
data.head(11)

children  days_employed  dob_years education  education_id  \
0          1   -8437.673028         42    высшее             0   
1          1   -4024.803754         36   среднее             1   
2          0   -5623.422610         33   Среднее             1   
3          3   -4124.747207         32   среднее             1   
4          0  340266.072047         53   среднее             1   
5          0    -926.185831         27    высшее             0   
6          0   -2879.202052         43    высшее             0   
7          0    -152.779569         50   СРЕДНЕЕ             1   
8          2   -6929.865299         35    ВЫСШЕЕ             0   
9          0   -2188.756445         41   среднее             1   
10         2   -4171.483647         36    высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   

     total_income                     purpose  
0   253875.639453               покупка жилья  
1   112080.014102     приобретение автомобиля  
2   145885.952297               покупка жилья  
3   267628.550329  дополнительное образование  
4   158616.077870             сыграть свадьбу  
5   255763.565419               покупка жилья  
6   240525.971920           операции с жильем  
7   135823.934197                 образование  
8    95856.832424       на проведение свадьбы  
9   144425.938277     покупка жилья для семьи  
10  113943.491460        покупка недвижимости

In [2]:
# вывод информации о размере таблицы
data.shape

(21525, 12)

In [3]:
#вывод общей информации о таблице (тип данных в столбце, кол-во записей в столбце)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Шаг 2. 
Предобработайте данные:

    определите и заполните пропущенные значения;
    замените вещественный тип данных на целочисленный;
    удалите дубликаты;
    выделите леммы в значениях столбца с целями получения кредита;
    категоризируйте данные.

Из общей информации видно, что пропуски есть только в столбцах 'days_employed' и 'total_income'.

In [4]:
# Количество пропусков в столбцах:
print("Количество пропусков в столбце 'days_employed' (трудовой стаж):", data['days_employed'].isna().sum())
print("Количество пропусков в столбце 'total_income' (доход в месяц):", data['total_income'].isna().sum())

Количество пропусков в столбце 'days_employed' (трудовой стаж): 2174
Количество пропусков в столбце 'total_income' (доход в месяц): 2174


In [5]:
# выделим строки с пропусками
data[data['days_employed'].isna() == True].head()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

# Комментарий наставника
Хорошая идея. Она кажется очевидной, но многие её почему-то пропускали. Значит, не такая она и очевидная, но очень правильная и хорошая

In [6]:
# проверим, есть ли строки с пропущенным стажем, но указанным доходом
data[(data['days_employed'].isna() == True) & (data['total_income'].isna() == False)]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

Таких строк нет (если не указан трудовой стаж, то не указана и зарплата), можно предположить что эти пропуски связаны, т.е. не случайны. Например, так могут вноситься клиенты банка, которые не подтвердили место работы или доход (устроенные неофициально).

# Комментарий наставника
Вот тут как-то непоследовательно (И это кажется единственная серьезная ошибка). 
days_employed ты заполняешь нулями. просто везде. а total_income как среднее тех кто с долгом, и тех кто без долга. 

Ошибок тут две:
1. Не очень правильно заполнять что-то исходя из основной исследуемой величины. Т.е. в идеале ты должен сделать вид, что не видишь кто вернул кредит, а кто нет, заполнить исходя из остальных данных пропуски и перейти к анализу. Дело в том, что это в некоторм смысле самосбывающийся прогноз: ты сначала исходя из наличия долга считаешь среднюю зп, а потом исходя из средней зп смотришь на то, как она влияет на долг. Это очевидно не честно. 
2. Мне не очень нравится, что days_employed = 0, а зп есть. Согласись это непоследовательно. Если ты предполагаешь наличие зарплаты, то и стаж какой-то должен быть. Ну или вообще откинуть столбец и не заполнять, если считаем, что он не нужен.

Плюс ещё такое замечание: 
1. Подумай исходя из чего логичнее заполнять заработную плату. Какая из имеющихся характеристик лучше всего должна коррелировать с total_income. Можно построить среднее значение total_income группируя по разным переменным (достаточно перебрать логичные варианты, а не вообще все) и посмотреть на результат.

В столбцах 'education', 'family_status', 'gender', 'income_type' содержатся данные, которые можно разделить на несколько категорий. Посмотрим, какие уникальные значения содержатся в каждом столбце.

In [7]:
# уникальные значения в столбце 'education' и их количество
data['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Выделяются несколько значений, которые отличаются только регистром. 

In [8]:
# приведем все символы к нижнему регистру
data['education'] = data['education'].str.lower()

# снова выведем список уникальных значений
data['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Остались только 5 категорий, которые отличаются друг от друга по смыслу.

Проверим остальные столбцы с категориальными данными.

In [9]:
data['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [10]:
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Даже если предположить, что XNA это не ошибка при вводе данных, а отдельная категория "неуказанный/невыбранный пол", удалим эту запись, т.к. 1 значение из ~21000 не значимо, но в дальнейшем может сделать анализ неудобным.


In [11]:
# оставим в датасете только те записи, у которых значение в столбце 'gender' не равно 'XNA'
data = data[data['gender'] != 'XNA']

# перенумеруем индексы
data = data.reset_index(drop = 'True')

# проверим список уникальных значений в столбце 'gender' и их количество
print(data['gender'].value_counts())

F    14236
M     7288
Name: gender, dtype: int64


In [12]:
data['income_type'].value_counts()

сотрудник          11119
компаньон           5084
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

Удалим записи со значениями 'предприниматель', 'безработный', 'студент','в декрете' в столбце 'income_type'. Количество записей мало, поэтому удаление не повлияет на итоговые выводы существенным образом.

In [13]:
# оставим в датасете только те записи, у которых значение в столбце 'income_type' не равно 
# 'предприниматель' или 'безработный' или 'студент' или 'в декрете'
data = data[(data['income_type'] != 'предприниматель') & \
      (data['income_type'] != 'безработный') & \
      (data['income_type'] != 'студент') & \
      (data['income_type'] != 'в декрете')]

# перенумеруем индексы
data = data.reset_index(drop = 'True')

# проверим список уникальных значений в столбце 'income_type' и их количество
print(data['income_type'].value_counts())

сотрудник      11119
компаньон       5084
пенсионер       3856
госслужащий     1459
Name: income_type, dtype: int64


Для заполнения пропущенных значений в столбцах 'days_employed' будем использовать средние значения трудового стажа для людей такого же возраста и такого же образования. 

Для этого создадим функцию, принимающую на вход значение возраста и категорию образования, и возвращающую среднее значение трудового стажа, подсчитанное для клиентов (с указанным трудовым стажем) с заданным возрастом и образованием.

In [14]:
def mean_days_employed(row):
    years = row['dob_years']
    education = row['education']
    return data[(data['days_employed'].isna() == False) \
              & (data['dob_years'] == years) \
              & (data['education'] == education)]['days_employed'].mean()

#### Дальше не смог применить функцию только к строкам с NaN, поэтому сделал присваивание через новые столбцы.

In [1]:
# создадим новый столбец, как результат применения функции к каждой строке в датасете
data['mean_days_employed'] = data.apply(mean_days_employed, axis = 1)

NameError: name 'data' is not defined

In [16]:
data.head(13)

children  days_employed  dob_years education  education_id  \
0          1   -8437.673028         42    высшее             0   
1          1   -4024.803754         36   среднее             1   
2          0   -5623.422610         33   среднее             1   
3          3   -4124.747207         32   среднее             1   
4          0  340266.072047         53   среднее             1   
5          0    -926.185831         27    высшее             0   
6          0   -2879.202052         43    высшее             0   
7          0    -152.779569         50   среднее             1   
8          2   -6929.865299         35    высшее             0   
9          0   -2188.756445         41   среднее             1   
10         2   -4171.483647         36    высшее             0   
11         0    -792.701887         40   среднее             1   
12         0            NaN         65   среднее             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   

     total_income                            purpose  mean_days_employed  
0   253875.639453                      покупка жилья         2374.265767  
1   112080.014102            приобретение автомобиля        -1263.661373  
2   145885.952297                      покупка жилья          339.142890  
3   267628.550329         дополнительное образование         1673.410241  
4   158616.077870                    сыграть свадьбу        91415.560024  
5   255763.565419                      покупка жилья        -1322.509097  
6   240525.971920                  операции с жильем         7070.099300  
7   135823.934197                        образование        49954.425498  
8    95856.832424              на проведение свадьбы        -2034.546455  
9   144425.938277            покупка жилья для семьи         2425.889224  
10  113943.491460               покупка недвижимости          791.306521  
11   77069.234271  покупка коммерческой недвижимости         4586.269530  
12            NaN                    сыграть свадьбу       296379.220657

In [17]:
# новый столбец, равный столбцу 'days_employed', но с заполненными пропусками 
data['days_employed_0'] = data['days_employed'].fillna(0)

# записываем значения столбца 'days_employed', как старые, в случае, если пропуска в столбце нет (.isna() возвращает 0)
# и как среднее значение трудового стажа для клиентов такого же возраста с таким же образованием, вместо пропуска
data['days_employed'] = \
data['days_employed_0'] + data['days_employed'].isna() * data['mean_days_employed']


In [18]:
data.head(13)

children  days_employed  dob_years education  education_id  \
0          1   -8437.673028         42    высшее             0   
1          1   -4024.803754         36   среднее             1   
2          0   -5623.422610         33   среднее             1   
3          3   -4124.747207         32   среднее             1   
4          0  340266.072047         53   среднее             1   
5          0    -926.185831         27    высшее             0   
6          0   -2879.202052         43    высшее             0   
7          0    -152.779569         50   среднее             1   
8          2   -6929.865299         35    высшее             0   
9          0   -2188.756445         41   среднее             1   
10         2   -4171.483647         36    высшее             0   
11         0    -792.701887         40   среднее             1   
12         0  296379.220657         65   среднее             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   

     total_income                            purpose  mean_days_employed  \
0   253875.639453                      покупка жилья         2374.265767   
1   112080.014102            приобретение автомобиля        -1263.661373   
2   145885.952297                      покупка жилья          339.142890   
3   267628.550329         дополнительное образование         1673.410241   
4   158616.077870                    сыграть свадьбу        91415.560024   
5   255763.565419                      покупка жилья        -1322.509097   
6   240525.971920                  операции с жильем         7070.099300   
7   135823.934197                        образование        49954.425498   
8    95856.832424              на проведение свадьбы        -2034.546455   
9   144425.938277            покупка жилья для семьи         2425.889224   
10  113943.491460               покупка недвижимости          791.306521   
11   77069.234271  покупка коммерческой недвижимости         4586.269530   
12            NaN                    сыграть свадьбу       296379.220657   

    days_employed_0  
0      -8437.673028  
1      -4024.803754  
2      -5623.422610  
3      -4124.747207  
4     340266.072047  
5       -926.185831  
6      -2879.202052  
7       -152.779569  
8      -6929.865299  
9      -2188.756445  
10     -4171.483647  
11      -792.701887  
12         0.000000

Для заполнения пропущенных значений в столбце 'total_income' выполним те же действия, что и в случае со столбцом 'days_employed'.

In [19]:
def mean_total_income(row):
    years = row['dob_years']
    education = row['education']
    return data[(data['total_income'].isna() == False) \
                & (data['dob_years'] == years) \
               & (data['education'] == education)]['total_income'].mean()

# создадим новый столбец, как результат применения функции к каждой строке в датасете
data['mean_total_income'] = data.apply(mean_total_income, axis = 1)

# новый столбец, равный столбцу 'total_income', но с заполненными пропусками 
data['total_income_0'] = data['total_income'].fillna(0)

# записываем значения столбца 'total_income', как старые, в случае, если пропуска в столбце нет (.isna() возвращает 0)
# и как среднее значение дохода в месяц для клиентов такого же возраста с таким же образованием, вместо пропуска
data['total_income'] = \
data['total_income_0'] + data['total_income'].isna() * data['mean_total_income']

In [20]:
data.head(13)

children  days_employed  dob_years education  education_id  \
0          1   -8437.673028         42    высшее             0   
1          1   -4024.803754         36   среднее             1   
2          0   -5623.422610         33   среднее             1   
3          3   -4124.747207         32   среднее             1   
4          0  340266.072047         53   среднее             1   
5          0    -926.185831         27    высшее             0   
6          0   -2879.202052         43    высшее             0   
7          0    -152.779569         50   среднее             1   
8          2   -6929.865299         35    высшее             0   
9          0   -2188.756445         41   среднее             1   
10         2   -4171.483647         36    высшее             0   
11         0    -792.701887         40   среднее             1   
12         0  296379.220657         65   среднее             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   

     total_income                            purpose  mean_days_employed  \
0   253875.639453                      покупка жилья         2374.265767   
1   112080.014102            приобретение автомобиля        -1263.661373   
2   145885.952297                      покупка жилья          339.142890   
3   267628.550329         дополнительное образование         1673.410241   
4   158616.077870                    сыграть свадьбу        91415.560024   
5   255763.565419                      покупка жилья        -1322.509097   
6   240525.971920                  операции с жильем         7070.099300   
7   135823.934197                        образование        49954.425498   
8    95856.832424              на проведение свадьбы        -2034.546455   
9   144425.938277            покупка жилья для семьи         2425.889224   
10  113943.491460               покупка недвижимости          791.306521   
11   77069.234271  покупка коммерческой недвижимости         4586.269530   
12  144255.121470                    сыграть свадьбу       296379.220657   

    days_employed_0  mean_total_income  total_income_0  
0      -8437.673028      243693.632170   253875.639453  
1      -4024.803754      161767.072002   112080.014102  
2      -5623.422610      168922.922098   145885.952297  
3      -4124.747207      148047.386144   267628.550329  
4     340266.072047      148317.316447   158616.077870  
5       -926.185831      204180.415473   255763.565419  
6      -2879.202052      245714.930658   240525.971920  
7       -152.779569      154577.479758   135823.934197  
8      -6929.865299      205725.048329    95856.832424  
9      -2188.756445      160935.019231   144425.938277  
10     -4171.483647      240352.009441   113943.491460  
11      -792.701887      165392.208314    77069.234271  
12         0.000000      144255.121470        0.000000

In [21]:
# удалим ненужные столбцы
del data['mean_days_employed']
del data['days_employed_0']
del data['mean_total_income']
del data['total_income_0']

In [24]:
# заменим тип данных в столбце 'total_income' на целочисленный методом astype('int')
data['total_income'] = data['total_income'].astype('int')

# заменим тип данных в столбце 'days_employed' на целочисленный методом astype('int')
data['days_employed'] = data['days_employed'].astype('int')

In [25]:
data.head(13)

children  days_employed  dob_years education  education_id  \
0          1          -8437         42    высшее             0   
1          1          -4024         36   среднее             1   
2          0          -5623         33   среднее             1   
3          3          -4124         32   среднее             1   
4          0         340266         53   среднее             1   
5          0           -926         27    высшее             0   
6          0          -2879         43    высшее             0   
7          0           -152         50   среднее             1   
8          2          -6929         35    высшее             0   
9          0          -2188         41   среднее             1   
10         2          -4171         36    высшее             0   
11         0           -792         40   среднее             1   
12         0         296379         65   среднее             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        144255   

                              purpose  
0                       покупка жилья  
1             приобретение автомобиля  
2                       покупка жилья  
3          дополнительное образование  
4                     сыграть свадьбу  
5                       покупка жилья  
6                   операции с жильем  
7                         образование  
8               на проведение свадьбы  
9             покупка жилья для семьи  
10               покупка недвижимости  
11  покупка коммерческой недвижимости  
12                    сыграть свадьбу

In [26]:
# выведем общую информацию о таблице
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21518 entries, 0 to 21517
Data columns (total 12 columns):
children            21518 non-null int64
days_employed       21518 non-null int64
dob_years           21518 non-null int64
education           21518 non-null object
education_id        21518 non-null int64
family_status       21518 non-null object
family_status_id    21518 non-null int64
gender              21518 non-null object
income_type         21518 non-null object
debt                21518 non-null int64
total_income        21518 non-null int64
purpose             21518 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


In [27]:
# выведем количество пропущенных значений в столбцах
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Все пропуски NaN заполнены. Числовые переменные записаны в виде целых чисел.

Проведем поиск дубликатов. Будем использовать метод duplicated(). 

In [28]:
# количество полных дубликатов
data.duplicated().sum()

71

# Комментарий наставника
Вот здесь я не соглашусь. У нас очень много одинаково заполненных полей и нет никаких уникальных цифр. Формулировки про цель кредита стандартные и часто повторяются. Ближе всего к уникальным только зарплаты и стаж, но их мы заполняли средними значениями почти для 2000 человек. Не удивительно, что где-то совпало. 

Я бы предложил вообще не делать drop_duplicates, если честно

Метод drop_duplicates() находит полные дубликаты - идентичные записи во всех столбцах. Однако, полностью одинаковые записи могут принадлежать разным людям, т.к. значения в некоторых столбцах - категория на выбор, а некоторые значения, которые могут быть действительно уникальными, ранее были заполнены средними значениями. Не будем проводить удаление дубликатов.

Проведем анализ столбца с целями кредитов.

In [29]:
# список уникальных записей в столбце 'purpose'
data['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     776
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   652
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    634
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      623
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              611
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Проведем лемматизацию записей в этом списке, разделим все цели получения кредита на несколько категорий и добавим к таблице столбец с указанием категории цели займа.

In [30]:
# импортируем библиотеку pymystem3 для лемматизации
from pymystem3 import Mystem
m = Mystem()

# проведем лемматизацию всех записей столбца с целями кредитов
purposes_lem = []
k = 0
for i in range(len(data)):
    lemmas = m.lemmatize(data['purpose'][i])
    purposes_lem = purposes_lem + lemmas

In [31]:
# импортируем контейнер Counter
from collections import Counter

# подсчет одинаковых значений и сортировка по частоте использования
Counter(purposes_lem).most_common()

[(' ', 33666),
 ('\n', 21518),
 ('недвижимость', 6364),
 ('покупка', 5909),
 ('жилье', 4471),
 ('автомобиль', 4314),
 ('образование', 4022),
 ('с', 2924),
 ('операция', 2610),
 ('свадьба', 2347),
 ('свой', 2235),
 ('на', 2232),
 ('строительство', 1880),
 ('высокий', 1375),
 ('получение', 1316),
 ('коммерческий', 1315),
 ('для', 1293),
 ('жилой', 1232),
 ('сделка', 944),
 ('дополнительный', 909),
 ('заниматься', 908),
 ('проведение', 776),
 ('сыграть', 774),
 ('сдача', 652),
 ('семья', 641),
 ('собственный', 634),
 ('со', 630),
 ('ремонт', 611),
 ('подержанный', 489),
 ('подержать', 479),
 ('приобретение', 462),
 ('профильный', 436)]

Анализируя этот набор лемм и уникальные значения для столбца 'purpose', выделим несколько категорий целей кредита:
    
    'жилая недвижимость'
    'коммерческая недвижимость'
    'автомобиль'
    'образование'
    'свадьба'.

Создадим новый столбец, в котором укажем одну из категорий целей кредита в зависимости от леммы слов в цели кредита.

In [32]:
# создадим функцию, присваивающую ту или иную категорию в зависимости от содержания столбца 'purpose'
def category(string):
    lemmas = m.lemmatize(string)
    for i in range(len(lemmas)):
        if (lemmas[i] == 'жилье') | (lemmas[i] == 'недвижимость'):
            return 'жилая недвижимость'
        if (lemmas[i] == 'коммерческий'):
            return 'коммерческая недвижимость'
        if (lemmas[i] == 'автомобиль'):
            return 'автомобиль'
        if (lemmas[i] == 'образование'):
            return 'образование'
    return 'свадьба'

In [33]:
# новый столбец, значения в котором - результат работы функции category получающей на вход значения в столбце 'purpose'
data['purpose_category'] = data['purpose'].apply(category)

In [34]:
# вывод первых 5 записей набора данных
data.head(5)

children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose    purpose_category  
0               покупка жилья  жилая недвижимость  
1     приобретение автомобиля          автомобиль  
2               покупка жилья  жилая недвижимость  
3  дополнительное образование         образование  
4             сыграть свадьбу             свадьба

Проверим результат работы функции, все ли цели верно распределены по категориям. Для этого создадим сводную таблицу.

In [35]:
# сводная таблица для анализа работы функции category, например выведем количество (функция count()) 
# записей в столбце 'debt', сгруппированных по стобцу 'gender', группировка данных при этом будет происходить 
# по столбцам 'purpose_category' и 'purpose'
data_pivot = data.pivot_table(index = ['purpose_category','purpose'], columns = 'gender', values = 'debt', aggfunc = 'count')
print(data_pivot)

gender                                                              F    M
purpose_category          purpose                                         
автомобиль                автомобили                              322  156
                          автомобиль                              338  156
                          на покупку автомобиля                   295  177
                          на покупку подержанного автомобиля      324  155
                          на покупку своего автомобиля            338  167
                          приобретение автомобиля                 320  142
                          свой автомобиль                         313  167
                          сделка с автомобилем                    299  156
                          сделка с подержанным автомобилем        319  170
жилая недвижимость        жилье                                   409  238
                          недвижимость                            435  199
                         

Видим, что все цели правильно распределены по выбранным категориям.

Разделим клиентов на 2 категории: с детьми и без. Добавим новый столбец 'children_type' в котором запишем значения 0, если нет детей и 1 если есть дети с помощью функции children_type(), принимающей на вход кол-во детей из столбца 'children'.

# комментарий наставника
Лучше, наверное, назвать with_children или has_children. Это более стандартное и распространнеое обозначение

In [36]:
def with_children(number_of_children):
    if number_of_children == 0:
        return 0
    return 1

# создадим новый столбец как результат работы функции children_type(), примененной к столбцу 'children'
data['with_children'] = data['children'].apply(with_children)

Проверим результат работы функции, для этого выведем сводную таблицу.

In [37]:
# сводная таблица: количество записей в столбце 'debt',
# сгруппированных по столбцу 'gender',
# группировка по столбцам 'children_type' и 'children'
data_pivot = data.pivot_table(index = ['with_children','children'], columns = 'gender', values = 'debt', aggfunc = 'count')
print(data_pivot)

gender                     F     M
with_children children            
0              0        9568  4576
1             -1          35    12
               1        3094  1723
               2        1258   796
               3         196   134
               4          28    13
               5           7     2
               20         47    29


Видим, что функция сработала правильно, клиентам без детей присвоен 0, клиентам с любым количеством детей присвоена 1.

Заметим, также, необычные значения в столбце 'children': -1 и 20, с достаточно большим числом записей. Предположим, что такая форма записи несет какой-то смысл, например -1 для клиентов, ожидающих ребенка, но ещё его не имеющего, а 20 для всех клиентов с количеством детей больше 5. Не станем удалять эти записи.

Рассмотрим доход в месяц. Узнаем минимальный и максимальный доход, затем попробуем выделить несколько категорий.

In [38]:
print(data['total_income'].min())
print(data['total_income'].max())

20667
2265604


Выделим 6 категорий дохода в месяц:

    '<100.000'
    '100.000 - 200.000'
    '200.000 - 350.000'
    '350.000 - 500.000'
    '500.000 - 1.000.000'
    '> 1.000.000'

# комментарий наставника
Хороший вопрос, как бить по доходу на группы. Идеального решения нет, можно и так, а можно посмотреть на разные статистики и выделить группы как-то иначе. Но это будет в следующих спринтах, как мне кажется, т.к. надо строить графики и т.д.

In [39]:
# создадим функцию, присваивающую ту или иную категорию в зависимости от содержания столбца 'total_income'
def category_income(string):
    if string < 100000:
        return '<100.000'
    if (string >= 100000) & (string < 200000):
        return '100.000 - 200.000'
    if (string >= 200000) & (string < 350000):
        return '200.000 - 350.000'
    if (string >= 350000) & (string < 500000):
        return '350.000 - 500.000'
    if (string >= 500000) & (string < 1000000):
        return '500.000 - 1.000.000'
    return '> 1.000.000'

In [40]:
# новый столбец, значения в котором - результат работы функции category_income, 
# получающей на вход значения в столбце 'total_income'
data['income_category'] = data['total_income'].apply(category_income)

In [41]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose    purpose_category  with_children  \
0               покупка жилья  жилая недвижимость              1   
1     приобретение автомобиля          автомобиль              1   
2               покупка жилья  жилая недвижимость              0   
3  дополнительное образование         образование              1   
4             сыграть свадьбу             свадьба              0   

     income_category  
0  200.000 - 350.000  
1  100.000 - 200.000  
2  100.000 - 200.000  
3  200.000 - 350.000  
4  100.000 - 200.000

In [42]:
# количество клиентов в категориях
data.by('income_category')['debt'].count()

income_category
100.000 - 200.000      11649
200.000 - 350.000       4552
350.000 - 500.000        635
500.000 - 1.000.000      197
<100.000                4460
> 1.000.000               25
Name: debt, dtype: int64

Количество записей в категориях существенно отличается, изменим границы категорий, чтобы добиться более равномерного распределения по категориям дохода.

In [43]:
def category_income_new(string):
    if string < 100000:
        return '<100.000'
    if (string >= 100000) & (string < 150000):
        return '100.000 - 150.000'
    if (string >= 150000) & (string < 200000):
        return '150.000 - 200.000'
    if (string >= 200000) & (string < 300000):
        return '200.000 - 300.000'
    return '> 300.000'

data['income_category'] = data['total_income'].apply(category_income_new)
print(data.by('income_category')['debt'].count())

income_category
100.000 - 150.000    6252
150.000 - 200.000    5397
200.000 - 300.000    3927
<100.000             4460
> 300.000            1482
Name: debt, dtype: int64


Имеем 4 группы дохода с сопоставимым количеством клиентов в каждой категории.

# Задание 2. 
Опишите свои выводы

## Шаг 1.
Ответьте на вопросы:

    Есть ли зависимость между наличием детей и возвратом кредита в срок?
    Есть ли зависимость между семейным положением и возвратом кредита в срок?
    Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
    Как разные цели кредита влияют на его возврат в срок?

Ответы сопроводите интерпретацией: поясните, о чём именно говорит полученный вами результат.


В этом разделе будем анализировать зависимость возврата кредита в срок от разных характеристик. Будем использовать группировку по какому-либо признаку и вычислять среднее значение столбца 'debt' для каждой выделенной группы.

В рамках этого исследования можно сделать предположение о том, что среднее значение для группы будет равно вероятности того, что новый клиент с таким же признаком не вернет кредит в срок (будет иметь задолженность). 

### Наличие детей
Проанализируем зависимость возврата кредита в срок от наличия детей. Сгруппируем датасет по столбцу 'with_children' и посчитаем среднее значение столбца 'debt' в полученных наборах. 

In [45]:
# сгруппируем по столбцу 'with_children', вычислим среднее значение столбца 'debt' для каждой группы, 
# отсортируем по убыванию
data.by('with_children')['debt'].mean().sort_values(ascending = False)

with_children
1    0.091673
0    0.075156
Name: debt, dtype: float64

Видим разные средние значения столбца 'debt' для разных групп. Так, для клиентов имеющих детей вероятность иметь задолженность - 9,2 %. Для клиентов без детей - 7.5%.

Можно сделать вывод, о том, что для банка безопаснее выдавать кредит клиентам, не имеющим детей.

### Семейное положение
Проанализируем зависимость возврата кредита в срок от семейного положения. Сгруппируем датасет по столбцу 'family_status' и посчитаем среднее значение столбца 'debt' в полученных наборах. 

In [46]:
# сгруппируем по столбцу 'family_status', вычислим среднее значение столбца 'debt' для каждой группы, 
# отсортируем по убыванию
data.by('family_status')['debt'].mean().sort_values(ascending = False)

family_status
Не женат / не замужем    0.097440
гражданский брак         0.092956
женат / замужем          0.075059
в разводе                0.071130
вдовец / вдова           0.065625
Name: debt, dtype: float64

Видим разные средние значения столбца 'debt' для разных групп. Так, для клиентов не состоящих в браке вероятность иметь задолженность - 9,7 %. Для клиентов в браке - 7.5%.

Можно сделать вывод, о том, что для банка безопаснее выдавать кредит клиентам в браке/разводе/овдовевшим, чем клиентам, не состоящим в браке.


### Суммарный доход
Проанализируем зависимость возврата кредита в срок от суммарного дохода в месяц. Сгруппируем датасет по столбцу 'income_category' и посчитаем среднее значение столбца 'debt' в полученных наборах. 

In [47]:
# сгруппируем по столбцу 'income_category', вычислим среднее значение столбца 'debt' для каждой группы, 
# отсортируем по убыванию
data.by('income_category')['debt'].mean().sort_values(ascending = False)

income_category
150.000 - 200.000    0.088753
100.000 - 150.000    0.085413
<100.000             0.078924
> 300.000            0.071525
200.000 - 300.000    0.068245
Name: debt, dtype: float64

Видим разные средние значения столбца 'debt' для разных групп. Так, для клиентов имеющих доход 150.000-200.000 вероятность иметь задолженность - 8,8 %. Для клиентов имеюищх доход 200.000-300.000 - 6.8%.

Можно сделать вывод, о том, что для банка безопаснее выдавать кредит клиентам с уровнем дохода 200.000-300.000.

### Цели кредита
Проанализируем зависимость возврата кредита в срок от цели кредита. Сгруппируем датасет по столбцу 'purpose_category' и посчитаем среднее значение столбца 'debt' в полученных наборах.

In [48]:
# сгруппируем по столбцу 'purpose_category', вычислим среднее значение столбца 'debt' для каждой группы, 
# отсортируем по убыванию
data.by('purpose_category')['debt'].mean().sort_values(ascending = False)

purpose_category
автомобиль                   0.093185
образование                  0.091994
свадьба                      0.079250
коммерческая недвижимость    0.075285
жилая недвижимость           0.071639
Name: debt, dtype: float64

Видим разные средние значения столбца 'debt' для разных групп. Так, для клиентов оформляющих кредит на автомобиль вероятность иметь задолженность - 9,3 %. Для клиентов оформляющих кредит на жилую недвижимость - 7.2%.

Можно сделать вывод, о том, что для банка безопаснее выдавать кредит на цели, связанные с недвижимостью.


# Задание 2. 
Опишите свои выводы

## Шаг 2.
Напишите общий вывод.


В ходе проведенного исследования, были получены средние значения вероятности задолженности по кредиту для разных групп клиентов. Были выявлены наиболее безопасные с точки зрения кредитования категории клиентов: без детей, состоящие в браке/разводе, с доходом в месяц 200.000-300.000, оформляющих кредит на цели, связанные с недвижимостью.

#### Коротко о причинах задержки: 
- поздно начал (к моменту начала курса ещё не закончил бесплатную вводную часть), поэтому частично знакомился с основными методами из введения уже во время курса;
- иногда много времени трачу на какие-то отдельные ситуации и не всегда могу их решить (из того, что помню - это поиск пропусков None по всей таблице с помощью двух циклов: по строкам и столбцам; и примение функции методом .apply только к некоторым строкам, для перезаписи значений (здесь хотел использовать это для заполнения пропусков в столбцах 'days_employed' и 'total_income'), в целом пока только привыкаю и знакомлюсь, не всегда бывает просто превратить последовательность действий, которые хочу сделать, в код, приходится возвращаться к тренажеру и шпаргалкам, но это, скорее, мелочи;
- иногда бывает непонятно, насколько хорошо/полно нужно сделать задачу, например в случае с заполнением пропусков, некоторые столбцы можно даже не рассматривать, потому что они не важны для итоговых выводов, можно заполнить средними значениями по 1-й совпадающей категории, или по 2-м, или как-то ещё по-другому; я понимаю, что это самостоятельная работа, и так и должно быть, но бывает, что это немного задерживает;
- внешние обстоятельства тоже, но в дальнейшем они будут мешать уже меньше, так что надеюсь догнать.